### PR TITLE
Fix/3362 grpc timestamp access

### PR DESCRIPTION
**Pull request by remo-lab**
## Describe the changes that are made
This PR fixes an issue where the replay service always accessed HTTPReq.Timestamp from the first test case, even when the first test case was a gRPC request. This resulted in incorrect or zero timestamps being used during mock matching.

To address this cleanly, the timestamp selection logic has been extracted into a dedicated helper function and both replay paths now use this unified function.

What was the problem?
When replaying test cases, the code assumed that the first entry in testCases was always an HTTP test. For gRPC-first scenarios, this caused:
an incorrect zero-value timestamp
inconsistent backdate calculations
potential nil dereferences
incorrect mock matching behavior
This bug is tracked in Issue #3362.

What’s changed?
1. Extracted getBackdateTimestamp helper (replay.go: 102–112):
A new reusable function now handles selecting the correct timestamp based on the test case kind (HTTP or GRPC_EXPORT).
This eliminates duplicated logic and ensures both replay paths use the same behavior.

2. Updated both replay paths to use the helper:
Docker Compose path (line ~747)
Non-Docker path (line ~838)
Both now simply call:
backdate := getBackdateTimestamp(testCases)
This guarantees consistent timestamp handling.

3. Added comprehensive unit tests (replay_test.go)
All tests now call the real getBackdateTimestamp() function — no duplicated logic.

The following cases are covered:
gRPC-first test case (original bug scenario)
HTTP-first test case (regression test)
Empty test case list
Nil test case entry

To run:
go test -v ./pkg/service/replay -run TestGetBackdateTimestamp
i have also provided the screenshot  ahter running test below.

## Links & References

Why this fix matters:
Ensures correct timestamps are used for both HTTP and gRPC test cases
Eliminates the previous zero-value timestamp bug
Removes duplicated logic and makes replay behavior consistent
Adds reliable test coverage for all edge cases
Improves maintainability and safety for future changes

Files Modified:
pkg/service/replay/replay.go (timestamp logic extracted + call sites updated)
pkg/service/replay/replay_test.go (new tests added)

**Closes:** #[issue number that will be closed through this PR]
- NA (if very small change like typo, linting, etc.)

### 🔗 Related PRs
- NA
### 🐞 Related Issues
- NA
### 📄 Related Documents
- NA

## What type of PR is this? (check all applicable)
- [ ] 📦 Chore
- [ ] 🍕 Feature
- [✔️ ] 🐞 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [✔️ ] ✅ Test
- [ ] 🔁 CI
- [ ] ⏩ Revert

## Added e2e test pipeline?
- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added comments for hard-to-understand areas?
- [ ] 👍 yes
- [ ] 🙅 no, because the code is self-explanatory

## Added to documentation?
- [ ] 📜 README.md
- [ ] 📓 Wiki
- [ ] 🙅 no documentation needed

## Are there any sample code or steps to test the changes?
- [✔️ ] 👍 yes, i have added a test file
- [ ] 🙅 no, because it is not needed

## Self Review done?
- [✔️ ] ✅ yes
- [ ] ❌ no, because I need help

## Any relevant screenshots, recordings or logs?
- 
Screenshot 2025-12-11 212328

## 🧠 Semantics for PR Title & Branch Name

Please ensure your PR title and branch name follow the Keploy semantics:

📌 [PR Semantics Guide](https://github.com/keploy/keploy/wiki/PR-Semantics)  
📌 [Branch Semantics Guide](https://github.com/keploy/keploy/wiki/Branch-Semantics)

**Examples:**

- **PR Title**: `fix: patch MongoDB document update bug`  
- **Branch Name**: `feat/#1-login-flow` (You may skip mentioning the issue number in the branch name if the change is small and the PR description clearly explains it.)

---

## Additional checklist:
- [✔️ ] Have you read the [Contributing Guidelines on issues](https://keploy.io/docs/keploy-explained/contribution-guide/)?
- [✔️ ] Have you followed the [PR Semantics guide](https://github.com/keploy/keploy/wiki/PR-Semantics) for naming this PR?
- [✔️ ] Have you followed the [Branch Semantics guide](https://github.com/keploy/keploy/wiki/Branch-Semantics) for naming your branch?

### CODE REVIEW

**Comment by AkashKumar7902** [comment]
@remo-lab please address copilot comments

File: pkg/service/replay/replay.go [hunk 1]
```diff
@@ -99,6 +99,18 @@ func NewReplayer(logger *zap.Logger, testDB TestDB, mockDB MockDB, reportDB Repo
...
+func getBackdateTimestamp(testCases []*models.TestCase) time.Time {
+	var backdate time.Time
+	if len(testCases) > 0 && testCases[0] != nil {
+		if testCases[0].Kind == models.HTTP {
+			backdate = testCases[0].HTTPReq.Timestamp
+		} else if testCases[0].Kind == models.GRPC_EXPORT {
+			backdate = testCases[0].GrpcReq.Timestamp
+		}
+	}
+	return backdate
```
Comment (AkashKumar7902) [inline pkg/service/replay/replay.go:126]: the branching is too much, can we shorten it? also what happens if backdate is returned as it is (means it doesn't go on the if condition of 104), shouldn't we log a warning ?

Comment (AkashKumar7902) [reply pkg/service/replay/replay.go:126]: also improve the function name, it should be short and to the point

**Review by AkashKumar7902 — COMMENTED** [review]
please address the review comments

**Comment by AkashKumar7902** [comment]
Pls link the issue corresponding to this pr

**Comment by remo-lab** [comment]
HI @AkashKumar7902 ,Thanks for the feedback! I’ve made the requested updates and refined the implementation:

Simplified the branching by switching to a switch statement, which removes the nested conditions and makes the logic easier to read.

1)Improved the function name to backdate so it’s shorter and more aligned with Go naming conventions.

2)Converted it into a method on Replayer so we can use the instance logger (r.logger) for warnings.

3)Added warning logs for all edge cases (empty test cases, unknown kind, zero timestamps) to make unexpected flows more visible.

4)Added a doc comment explaining what the method does and how it handles HTTP and gRPC test cases.

5)Updated the tests to use r.backdate() directly and removed the trailing newline that Copilot flagged.

Everything compiles cleanly and all tests are passing. Let me know if you’d like anything else improved — happy to adjust!

**Comment by remo-lab** [comment]
Hi @AkashKumar7902 ,I have Implemented the suggested changes for this PR.”

**Comment by remo-lab** [comment]
Hi @AkashKumar7902, I’ve implemented the suggested changes for this PR.  
Whenever you get time, could you please review it again? Thanks!

**Comment by remo-lab** [comment]
Hi @Sarthak160 @gouravkrosx , can you please approve this pr?

**Comment by remo-lab** [comment]
HI @AkashKumar7902 @gouravkrosx @Sarthak160 could you please approve this pr?

**Comment by remo-lab** [comment]
Hi @gouravkrosx @Sarthak160 does this pr need any further update?
